File: geekbang/algorithm-models/3.5-子集递归模板/subsets.go
package main

func subsets(nums []int) [][]int {
	var s []int
	var ans [][]int
	var findSubsets func(index int)
	findSubsets = func(index int) {
		if index == len(nums) {
			ans = append(ans, make([]int, 0))
			for _, i := range s {
				ans[len(ans)-1] = append(ans[len(ans)-1], i)
			}
			return
		}
		findSubsets(index + 1)
		s = append(s, nums[index])
		findSubsets(index + 1)
		s = s[:len(s)-1]
	}
	findSubsets(0)
	return ans
}

File: geekbang/algorithm-models/3.5-子集递归模板/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.ans = []
        self.s = []
        self.n = len(nums)
        self.findSubsets(nums, 0)
        return self.ans

    def findSubsets(self, nums, index):
        if index == self.n:
            self.ans.append(self.s[:]) # make a copy
            return
        self.findSubsets(nums, index + 1)
        self.s.append(nums[index])
        self.findSubsets(nums, index + 1)
        self.s.pop()


